.github/gh-actions-self-hosted-runners/arc/images/Dockerfile

```diff
@@ -41,6 +41,9 @@ RUN curl -OL https://cdn.azul.com/zulu/bin/zulu8.70.0.23-ca-jdk8.0.372-linux_x64
     tar -C /usr/local -xzf zulu8.70.0.23-ca-jdk8.0.372-linux_x64.tar.gz && \
     rm zulu8.70.0.23-ca-jdk8.0.372-linux_x64.tar.gz && \
     mv /usr/local/zulu8.70.0.23-ca-jdk8.0.372-linux_x64 /usr/local/java
+RUN curl -OL https://cdn.azul.com/zulu/bin/zulu17.46.19-ca-jdk17.0.9-linux_x64.tar.gz && \
+    tar -C /usr/local -xzf https://cdn.azul.com/zulu/bin/zulu17.46.19-ca-jdk17.0.9-linux_x64.tar.gz && \
+    rm zulu17.46.19-ca-jdk17.0.9-linux_x64.tar.gz
```

Comment:
Do we need something like ` mv /usr/local/zulu8.70.0.23-ca-jdk8.0.372-linux_x64 /usr/local/java` above in this line as well so that this is available on `PATH` rather than the current directory?

Reply:
I'm not sure how this works, but this is a thing:

```
% ./gradlew -q javaToolchains

 + Options
     | Auto-detection:     Enabled
     | Auto-download:      Enabled

 + Azul Zulu JDK 11.0.21+9-LTS
     | Location:           /Library/Java/JavaVirtualMachines/zulu-11.jdk/Contents/Home
     | Language Version:   11
     | Vendor:             Azul Zulu
     | Architecture:       x86_64
     | Is JDK:             true
     | Detected by:        MacOS java_home

 + Azul Zulu JDK 20.0.2+9
     | Location:           /Library/Java/JavaVirtualMachines/zulu-20.jdk/Contents/Home
     | Language Version:   20
     | Vendor:             Azul Zulu
     | Architecture:       x86_64
     | Is JDK:             true
     | Detected by:        MacOS java_home

 + Azul Zulu JDK 21.0.1+12-LTS
     | Location:           /Library/Java/JavaVirtualMachines/zulu-21.jdk/Contents/Home
     | Language Version:   21
     | Vendor:             Azul Zulu
     | Architecture:       x86_64
     | Is JDK:             true
     | Detected by:        Current JVM

 + Oracle JDK 1.8.0_221-b11
     | Location:           /Library/Java/JavaVirtualMachines/jdk1.8.0_221.jdk/Contents/Home
     | Language Version:   8
     | Vendor:             Oracle
     | Architecture:       x86_64
     | Is JDK:             true
     | Detected by:        MacOS java_home

 + Oracle JRE 1.8.0_221-b11
     | Location:           /Library/Internet Plug-Ins/JavaAppletPlugin.plugin/Contents/Home
     | Language Version:   8
     | Vendor:             Oracle
     | Architecture:       x86_64
     | Is JDK:             false
     | Detected by:        MacOS java_home
```

Which makes me think it may be that we just select them directly. Or just set JAVA_HOME directly to wherever we unpack it, etc.

Reply:
I know setup-java will setup toolchains like this. I'd expect if Java was already on the path it wouldn't reinstall it (though I don't know for sure) and would just set up the toolchain. You could take a look at the current java 8 runs in the setup-java step and see what's happening. Throwing it into a random folder probably will not help us by itself though.

Reply:
I don't have strong opinions on that. Mostly I expect that we should set up the java version explicitly in all cases. Maybe there's a "normal" way to set things up so that by specifying version "17" you get it to resolve to `zulu17.46.19-ca-jdk17.0.9`

Reply:
> I know setup-java will setup toolchains like this. I'd expect if Java was already on the path it wouldn't reinstall it (though I don't know for sure) and would just set up the toolchain.

Currently the java environment is an optional parameter of the setup-environment-action action. If not provided, the action runs on this pre-installed java version, e.g. https://github.com/apache/beam/blob/79b9de283feda3c4efea7be4e90c2b7ef4ebfb40/.github/workflows/beam_PreCommit_Java.yml#L178.  If provided, it will use the (back of the list) configured java version as the default java command, e.g.: https://github.com/apache/beam/blob/79b9de283feda3c4efea7be4e90c2b7ef4ebfb40/.github/workflows/beam_PreCommit_SQL_Java11.yml#L84

Reply:
So really the main thing is that this is just a pre-caching of whatever the result of this would be:

```
java-versions:
 - 8
 -  17
```

(making up yaml but I know we do have tests that specify having both installed)

Reply:
Yep, it will install both and then set the highest version as the default (if you don't specify a version). I'm still pretty confident that it needs to end up on the path though in order for it to get picked up.

Reply:
This looks to be potentially a totally different distribution? https://github.com/apache/beam/blob/79b9de283feda3c4efea7be4e90c2b7ef4ebfb40/.github/actions/setup-environment-action/action.yml#L51

I presume that action can be invoked repeatedly with multiple versions and they each get installed into some non-overlapping places.

Reply:
Which is to say that what I'm trying to do on the image is just make it idempotent w.r.t. `actions/setup-java` and I would love your advice on the best way to do that. Probably having this script in concordance with `actions/setup-java` is more of a goal than having it match the code it is copy/pasted from.

Reply:
> I presume that action can be invoked repeatedly with multiple versions and they each get installed into some non-overlapping places.

Correct

> Probably having this script in concordance with actions/setup-java is more of a goal than having it match the code it is copy/pasted from.

Yeah, that's right. Looking at https://github.com/actions/setup-java/blob/7a445ee88d4e23b52c33fdc7601e40278616c7f8/src/setup-java.ts#L71 the key thing is can we find it in the tool-cache - https://github.com/actions/setup-java/blob/7a445ee88d4e23b52c33fdc7601e40278616c7f8/src/distributions/base-installer.ts#L49

We're not actually doing that correctly with our other versions either. Looking at https://github.com/apache/beam/actions/runs/7585699321/job/20662163515 maybe it just doesn't matter though. It takes all of 6 seconds to download/extract Java 11. So I guess we don't need this, the primary value here is just making sure we have _some_ ok default version of Java downloaded + available on the path so that every job doesn't need to specify a java-version.

I guess my take is that this just isn't actually worth it until we commit to making Java 17 the default, at which point we should just replace the java 8 download/extraction

Reply:
FWIW I expected this to have a bigger impact on build times, but it just doesn't appear to be the case.

Reply:
OK if it only takes 6 seconds then this isn't a prerequisite for moving ahead with version upgrades.

To be clear, this was a step in the process of migrating to 17 as the default for CI, and from there bumping up the minimum version for build so that we could keep up with build tools that have dropped Java 8 support. Aka I think we can commit.